src/scalar_weighting.rs

```diff
@@ -0,0 +1,232 @@
+use std::borrow::Borrow;
+
+use tch::nn::{Init, Linear, Module, Path};
+use tch::{Kind, Reduction, Tensor};
+
+use crate::cow::CowTensor;
+use crate::traits::LayerOutput;
+
+/// Layer that performs a scalar weighting of layers.
+///
+/// Following Peters et al., 2018 and Kondratyuk & Straka, 2019, this
+/// layer applies scalar weighting:
+///
+/// *e = c ∑_i[ h_i · softmax(w)_i ]*
+#[derive(Debug)]
+pub struct ScalarWeight {
+    /// Layer dropout probability.
+    layer_dropout_prob: f64,
+
+    /// Layer-wise weights.
+    layer_weights: Tensor,
+
+    /// Scalar weight.
+    scale: Tensor,
+}
+
+impl ScalarWeight {
+    pub fn new<'a>(vs: impl Borrow<Path<'a>>, n_layers: i64, layer_dropout_prob: f64) -> Self {
+        assert!(
+            n_layers > 0,
+            "Number of layers ({}) should be larger than 0",
+            n_layers
+        );
+
+        assert!(
+            layer_dropout_prob >= 0.0 && layer_dropout_prob < 1.0,
+            "Layer dropout should be in [0,1), was: {}",
+            layer_dropout_prob
+        );
+
+        let vs = vs.borrow();
+
+        ScalarWeight {
+            layer_dropout_prob,
+            layer_weights: vs.var("layer_weights", &[n_layers], Init::KaimingUniform),
+            scale: vs.var("scale", &[], Init::Const(1.)),
+        }
+    }
+
+    pub fn forward(&self, layers: &[impl LayerOutput], train: bool) -> Tensor {
+        assert_eq!(
+            self.layer_weights.size()[0],
+            layers.len() as i64,
+            "Expected {} layers, got {}",
+            self.layer_weights.size()[0],
+            layers.len()
+        );
+
+        let layers = layers
+            .iter()
+            .map(LayerOutput::layer_output)
+            .collect::<Vec<_>>();
+
+        // Each layer has shape:
+        // [batch_size, sequence_len, layer_size],
+        //
+        // stack the layers to get a single tensor of shape:
+        // [batch_size, sequence_len, n_layers, layer_size]
+        let layers = Tensor::stack(&layers, 2);
```

Comment:
How does stack compare to a for loop?

Reply:
Good question! In the big picture, these differences will not be visible, since the encoder dominates things time-wise. But we could benchmark this in isolation. I will create an issue to write a criterion bench for this. We could then replace it by a for-loop if it is faster.